src/Microsoft.Owin.Security.OpenIdConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -308,17 +344,99 @@ protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
                     nonce = RetrieveNonce(openIdConnectMessage);
                 }
 
-                // remember 'session_state' and 'check_session_iframe'
-                if (!string.IsNullOrWhiteSpace(openIdConnectMessage.SessionState))
+                Options.ProtocolValidator.ValidateAuthenticationResponse(new OpenIdConnectProtocolValidationContext()
                 {
-                    ticket.Properties.Dictionary[OpenIdConnectSessionProperties.SessionState] = openIdConnectMessage.SessionState;
+                    ClientId = Options.ClientId,
+                    ProtocolMessage = openIdConnectMessage,
+                    ValidatedIdToken = jwt,
+                    Nonce = nonce
+                });
+
+                // Authorization Code Flow
+                if (openIdConnectMessage.Code != null && string.IsNullOrWhiteSpace(openIdConnectMessage.IdToken))
```

Comment:
Code will be populated in a hybrid request and we only want to execute this logic on 'code flow' so we need to check if the IdToken is empty to know that is the case.

Reply:
I don't want the code to diverge from Core here, it will be harder to maintain. I'd rather have an additional option like RedemeCode to prevent the new code from running for existing users.

Reply:
I agree that the behavior should not change for existing users and this code will not run unless they change the ResponseType to 'code' (code flow). This would cause the code to return null before this update so I don't expect this setting is being used. So the user already has to explicitly opt in to this new behavior by setting that parameter . Also, I did not see a RedemeCode option it the Core implementation.

Reply:
I'm also concerned about the code layout and flow diverging from Core. Processing things in a different order and with different conditions makes it harder to ensure we get the same result in both frameworks.

I'm asking you to use the same layout and flow from Core and mitigate the behavior changes for existing users with a new option RedemeCode.

Reply:
Chris, I’m working through this and I see that in Core the AuthorizationCodeReceivedContext provides a ‘HandledCodeRedemption’ and a ‘TokenEndpointResponse’ property that allows the user to handle the code redemption. Do you want this ported over? Also, this is what Core is using to determine if it should redeem the code, but it looks like the default is to run the code redemption no matter what the flow. It seems that this would be unnecessary for the Hybrid flow.

https://github.com/aspnet/AspNetCore/blob/e7398ca25d53432eb96dc7d4e0d56be501b771a4/src/Security/Authentication/OpenIdConnect/src/Events/AuthorizationCodeReceivedContext.cs#L55

https://github.com/aspnet/AspNetCore/blob/e7398ca25d53432eb96dc7d4e0d56be501b771a4/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L655

Reply:
> AuthorizationCodeReceivedContext provides a ‘HandledCodeRedemption’ and a ‘TokenEndpointResponse’ property that allows the user to handle the code redemption. Do you want this ported over?

Yes, if possible.

> Also, this is what Core is using to determine if it should redeem the code, but it looks like the default is to run the code redemption no matter what the flow. It seems that this would be unnecessary for the Hybrid flow.

Yes Core redeems the code by default, that's why I suggested a new option in Katana to opt in to redemption.

> It seems that this would be unnecessary for the Hybrid flow.

Some servers return minimized tokens on the front channel and fatter tokens on the backchannel. For those servers you need to redeem the code to get the rest of the user information.